store/etcd/put.go

```diff
@@ -3,236 +3,156 @@ package etcd
 	"github.com/segmentio/ksuid"
 )
 
-// Put adds a version of the given ruleset using an uuid.
+// Put stores the given rules under the rules tree. If no signature is found for the given path it returns an error.
 func (s *RulesetService) Put(ctx context.Context, path string, ruleset *regula.Ruleset) (*store.RulesetEntry, error) {
-	sig, err := validateRuleset(path, ruleset)
-	if err != nil {
-		return nil, err
-	}
-
-	var entry store.RulesetEntry
+	var entry *store.RulesetEntry
+	var err error
 
 	txfn := func(stm concurrency.STM) error {
-		// generate a checksum from the ruleset for comparison purpose
-		h := md5.New()
-		err = json.NewEncoder(h).Encode(ruleset)
-		if err != nil {
-			return errors.Wrap(err, "failed to generate checksum")
-		}
-		checksum := string(h.Sum(nil))
-
-		// if nothing changed return latest ruleset
-		if stm.Get(s.checksumsPath(path)) == checksum {
-			v := stm.Get(stm.Get(s.latestRulesetPath(path)))
-
-			var pbrse pb.RulesetEntry
-			err = proto.Unmarshal([]byte(v), &pbrse)
-			if err != nil {
-				s.Logger.Debug().Err(err).Str("entry", v).Msg("put: entry unmarshalling failed")
-				return errors.Wrap(err, "failed to unmarshal entry")
-			}
-
-			s.Logger.Debug().Str("path", path).Msg("ruleset didn't change, returning without creating a new version")
-
-			entry.Path = pbrse.Path
-			entry.Version = pbrse.Version
-			entry.Ruleset = rulesetFromProtobuf(pbrse.Ruleset)
-
-			return store.ErrNotModified
-		}
-
-		// make sure signature didn't change
-		rawSig := stm.Get(s.signaturesPath(path))
-		if rawSig != "" {
-			var pbsig pb.Signature
-
-			err := proto.Unmarshal([]byte(rawSig), &pbsig)
-			if err != nil {
-				s.Logger.Debug().Err(err).Str("signature", rawSig).Msg("put: signature unmarshalling failed")
-				return errors.Wrap(err, "failed to decode ruleset signature")
-			}
-
-			err = compareSignature(signatureFromProtobuf(&pbsig), sig)
-			if err != nil {
-				return err
-			}
-		}
-
-		// if no signature found, create one
-		if rawSig == "" {
-			b, err := proto.Marshal(signatureToProtobuf(sig))
-			if err != nil {
-				return errors.Wrap(err, "failed to encode updated signature")
-			}
-
-			stm.Put(s.signaturesPath(path), string(b))
-		}
+		p := rulesPutter{s, stm}
+		entry, err = p.put(ctx, path, ruleset)
+		return err
+	}
 
-		// update checksum
-		stm.Put(s.checksumsPath(path), checksum)
+	_, err = concurrency.NewSTM(s.Client, txfn, concurrency.WithAbortContext(ctx))
+	if err != nil && err != store.ErrRulesetNotModified && !store.IsValidationError(err) {
+		return nil, errors.Wrap(err, "failed to put ruleset")
+	}
 
-		// create a new ruleset version
-		k, err := ksuid.NewRandom()
-		if err != nil {
-			return errors.Wrap(err, "failed to generate ruleset version")
-		}
-		version := k.String()
+	return entry, err
+}
 
-		err = s.putVersions(stm, path, version)
-		if err != nil {
-			return err
-		}
+// rulesPutter is responsible for validating and storing rules, updating checksums and other actions
```

Comment:
:tulip: I like the `rulesPutter`, but I prefer to use the `rulesWood` or the `rules9Iron`.  :golf: